cpp/src/tsne/bh_kernels.cuh

```diff
@@ -635,17 +655,37 @@ __global__ void attractive_kernel_bh(
   const float *restrict VAL, const int *restrict COL, const int *restrict ROW,
   const float *restrict Y1, const float *restrict Y2,
   const float *restrict norm, float *restrict attract1,
-  float *restrict attract2, const int NNZ) {
+  float *restrict attract2, const int NNZ,
+  int *restrict flag_unstable_computation) {
   const int index = (blockIdx.x * blockDim.x) + threadIdx.x;
   if (index >= NNZ) return;
   const int i = ROW[index];
   const int j = COL[index];
 
   // TODO: Calculate Kullback-Leibler divergence
   // TODO: Convert attractive forces to CSR format
-  const float PQ = __fdividef(
-    VAL[index],
-    norm[i] + 1.0f + norm[j] - 2.0f * (Y1[i] * Y1[j] + Y2[i] * Y2[j]));  // P*Q
+  // Try single precision compute first
+  float denominator = __fmaf_rn(-2.0f, (Y1[i] * Y1[j]), norm[i] + 1.0f) +
+                      __fmaf_rn(-2.0f, (Y2[i] * Y2[j]), norm[j]);
+
+  if (__builtin_expect(denominator == 0, false)) {
+    double _Y1 = static_cast<double>(Y1[i] * Y1[j]);
+    double _Y2 = static_cast<double>(Y2[i] * Y2[j]);
```

Comment:
This still does the calculation with single-precision. At least one of the operands needs to be cast to double before the multiplication.

Reply:
Will update to cast both variables before the multiply.

The second code link simply fails the entire process when the cell race condition is hit. 
```cpp
            cell = atomicSub((int*)&bottomd, 1) - 1;
            if (cell <= nbodiesd) {printf("ERROR: out of cell memory\n"); asm("trap;");}
```
This PR leverages the fallback path for closely packed points that exceed the minimum radius.

It also looks like they've moved to an oct-tree formulation instead of the current quad tree. Definitely worth looking at, but it will need to be a later update.